lib/BridgedRoom.js

```diff
@@ -0,0 +1,123 @@
+"use strict";
+
+function BridgedRoom(bridge, gitter, gitterRealtime, matrixRoomModel, gitterRoomModel) {
+  this._bridge = bridge;
+  this._gitter = gitter;
+  this._gitterRealtime = gitterRealtime;
+  this._matrixRoomModel = matrixRoomModel;
+  this._gitterRoomModel = gitterRoomModel;
+}
+
+BridgedRoom.prototype.matrixRoomId = function() {
+  return this._matrixRoomModel.getId();
+};
+
+BridgedRoom.prototype.gitterRoomName = function() {
+  return this._gitterRoomModel.getId();
+};
+
+BridgedRoom.prototype.joinAndStart = function() {
+  // We have to find out our own gitter user ID so we can ignore reflections of
+  // messages we sent
+  return this._bridge.getMyGitterUserId().then((gitterUserId) => {
+    this._gitterUserId = gitterUserId;
+
+    return this._gitter.rooms.join(this.gitterRoomName()).then((room) => {
+      this._gitterRoom = room;
+
+      var events = room.streaming().chatMessages();
+
+      events.on('chatMessages', (message) => {
+        if (!message.model.fromUser || message.model.fromUser.id == gitterUserId) {
+          // Ignore a reflection of my own messages
+          return;
+        }
+
+        if (message.operation !== 'create' && message.operation !== 'update') {
+          return;
+        }
+
+        this.onGitterMessage(message);
+      });
+
+      this._gitterRealtime.subscribe("/v1/rooms/" + room.id, (message) => {
+        this._bridge.getGitterUserById(message.userId).then((user) => {
+          if (user)
+            user.setRoomPresence(room.id, message.status == 'in');
+        });
+      });
+    });
+  });
+};
+
+BridgedRoom.prototype.stopAndLeave = function() {
+  this._gitterRoom.streaming().disconnect();
+
+  return this._gitterRoom.removeUser(this._gitterUserId);
+};
+
+function quotemeta(s) { return s.replace(/\W/g, '\\$&'); }
+
+BridgedRoom.prototype.onGitterMessage = function(message) {
+  var fromUser = message.model.fromUser;
+
+  console.log('gitter->' + this.gitterRoomName() + ' from ' + fromUser.username + ':', message.model.text)
+
+  this._bridge.mapGitterUser(fromUser).then((user) => {
+    return user.update(fromUser)
+      .catch((e) => {
+        console.log("Updating user failed:", e);
+        // ignore the failure and continue anyway
+      }).then(() => {
+        return user;
+      });
+  }).then((user) => {
+    var matrixMessage = {
+      msgtype: "m.text",
+      body: message.model.text,
```

Comment:
It looks like this function is the main bit where gitter messages get mapped to matrix messages. It would be great if you could block-comment what `message` should look like, in order to aid readability:
```javascript
// Gitter messages look like:
// {
//    model: {
//      text: "plaintext",
//      html: "html text",
//      status: ???
//    }
//  }
```